dbms/src/Storages/Page/V3/LogFile/LogWriter.cpp

```diff
@@ -92,23 +93,54 @@ void LogWriter::close()
-    const int header_size = recycle_log_files ? Format::RECYCLABLE_HEADER_SIZE : Format::HEADER_SIZE;
-
-    // Fragment the record if necessary and emit it. Note that if payload is empty,
-    // we still want to iterate once to emit a single zero-length record.
-    bool begin = true;
-    size_t payload_left = payload_size;
-
-    size_t head_sizes = ((payload_size / Format::BLOCK_SIZE) + 1) * Format::RECYCLABLE_HEADER_SIZE;
-    if (payload_size + head_sizes >= buffer_size)
+    const UInt32 header_size = recycle_log_files ? Format::RECYCLABLE_HEADER_SIZE : Format::HEADER_SIZE;
+
+    // Check whether the payload can fit in current block.
+    // If not, need to fragment the record and emit it.
+    // Note that if payload is empty, we still want to iterate once to emit a single zero-length record.
+    size_t total_write_size = 0;
+    assert(block_offset <= Format::BLOCK_SIZE);
+    size_t current_block_avail_size = Format::BLOCK_SIZE - block_offset;
+    if (payload_size + header_size <= current_block_avail_size)
     {
-        size_t new_buff_size = payload_size + ((head_sizes / Format::BLOCK_SIZE) + 1) * Format::BLOCK_SIZE;
+        total_write_size = payload_size + header_size;
+    }
+    else
+    {
+        auto current_avail_payload_space = header_size <= current_block_avail_size ? (current_block_avail_size - header_size) : 0;
+        auto complete_fragment_num = (payload_size - current_avail_payload_space) / (Format::BLOCK_SIZE - header_size);
+        // Fill current block
+        total_write_size += current_block_avail_size;
+        // Fill `complete_fragment_num` blocks
+        total_write_size += complete_fragment_num * Format::BLOCK_SIZE;
+        // Fill a partial tail block if needed
+        size_t remaining_payload_size = payload_size - current_avail_payload_space - (Format::BLOCK_SIZE - header_size) * complete_fragment_num;
+        assert(payload_size >= current_avail_payload_space + (Format::BLOCK_SIZE - header_size) * complete_fragment_num);
+        assert(remaining_payload_size + header_size <= Format::BLOCK_SIZE);
+        total_write_size += (remaining_payload_size > 0) ? (remaining_payload_size + header_size) : 0;
+    }
+    assert(total_write_size > 0);
 
+    size_t buffer_avail_size = buffer_size - write_buffer.offset();
+    if (total_write_size > buffer_avail_size)
+    {
+        // Flush the buffer if it has data before reset buffer
+        if (unlikely(write_buffer.offset() > 0))
+        {
+            flush(write_limiter, /* background */ false);
+        }
+        assert(write_buffer.offset() == 0);
+        // Accurate new_buff_size should be ((total_write_size - 1) / Format::BLOCK_SIZE + 1) * Format::BLOCK_SIZE
+        // Just be simple and conservative here.
+        size_t new_buff_size = (total_write_size / Format::BLOCK_SIZE + 1) * Format::BLOCK_SIZE;
         buffer = static_cast<char *>(realloc(buffer, buffer_size, new_buff_size));
```

Comment:
Do not realloc the buffer size, keep the buffer size stick to `Format::BLOCK_SIZE`. And call `flush` after we finish writing a block.
The places we need to call `flush` should be limited.

This should be the correct way to fix this misbehavior.

Reply:
Fixed.